From bf0fae2efae704f7cf670a6cb844d2330ae02846 Mon Sep 17 00:00:00 2001
From: Andrey Lebedev <andrey@lebedev.lt>
Date: Thu, 9 Apr 2020 19:13:54 +0300
Subject: [PATCH] Fix software display rotation

Initial fix (8bd49afa25e112f3d5f377dfa1ad72656effc35d) caused rendering
problems (black screen) on platforms in which SOFTWARE_ROTATE_DISPLAY
is not defined (e.g. Android). It was reverted in
6ef5768a073f7fd07442dc6f897f4fe7220d460e.

Fixes #351
---
 src/Screen/OpenGL/Init.cpp | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/Screen/OpenGL/Init.cpp b/src/Screen/OpenGL/Init.cpp
index 7695359fdb..3129911675 100644
--- a/src/Screen/OpenGL/Init.cpp
+++ b/src/Screen/OpenGL/Init.cpp
@@ -322,12 +322,15 @@ OpenGL::SetupViewport(UnsignedPoint2D size)
   glViewport(0, 0, size.x, size.y);
 
 #ifdef SOFTWARE_ROTATE_DISPLAY
-  projection_matrix = glm::rotate(glm::mat4(1),
-                                  (GLfloat)Angle::Degrees(OrientationToRotation(display_orientation)).Radians(),
-                                  glm::vec3(0, 0, 1));
+  glm::mat4 rot_matrix = glm::rotate(
+    glm::mat4(1),
+    (GLfloat)Angle::Degrees(OrientationToRotation(display_orientation)).Radians(),
+    glm::vec3(0, 0, 1));
   OrientationSwap(size, display_orientation);
+#else
+  glm::mat4 rot_matrix = glm::mat4(1);  // no rotation
 #endif
-  projection_matrix = glm::ortho<float>(0, size.x, size.y, 0, -1, 1);
+  projection_matrix = rot_matrix * glm::ortho<float>(0, size.x, size.y, 0, -1, 1);
   UpdateShaderProjectionMatrix();
 
   viewport_size = size;
-- 
2.20.1

